### PR TITLE
Plugin/Kubernetes: Service and Endpoint Indexing

### PR DESCRIPTION
**Pull request by rajansandeep**
### 1. What does this pull request do?
Add service and endpoint indexing

### 2. Which issues (if any) are related?
#814 

### 3. Which documentation changes (if any) need to be made?
None

### CODE REVIEW

File: plugin/kubernetes/controller.go [hunk 1]
```diff
@@ -21,10 +21,18 @@ var (
 )
 
 const podIPIndex = "PodIP"
+const svcNameNamespaceIndex = "NameNamespace"
```
Comment (miekg) [inline plugin/kubernetes/controller.go:24]: More Go-like is to group these:
```
const (
   ...
)
```

File: plugin/kubernetes/controller.go [hunk 2]
```diff
@@ -110,6 +119,38 @@ func podIPIndexFunc(obj interface{}) ([]string, error) {
 	return []string{p.Status.PodIP}, nil
 }
 
+func svcIPIndexFunc(obj interface{}) ([]string, error) {
```
Comment (miekg) [inline plugin/kubernetes/controller.go:122]: Do these really need to take an empty interface?

File: plugin/kubernetes/controller.go [hunk 3]
```diff
@@ -255,6 +296,79 @@ func (dns *dnsControl) PodIndex(ip string) (pods []*api.Pod) {
...
+func (dns *dnsControl) EpIndex(idx string) (ep []*api.Endpoints) {
+	if dns.epLister == nil {
+		return nil
+	}
+	os, err := dns.epLister.ByIndex(epNameNamespaceIndex, idx)
+	if err != nil {
+		return nil
+	}
+	for _, o := range os {
+		e, ok := o.(*api.Endpoints)
```
Comment (miekg) [inline plugin/kubernetes/controller.go:345]: Are all these type checks really needed?

File: plugin/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -342,8 +342,18 @@ func (k *Kubernetes) findPods(r recordRequest, zone string) (pods []msg.Service,
...
-	for _, svc := range k.APIConn.ServiceList() {
+	var (
+		endpointsList []*api.Endpoints
+		serviceList   []*api.Service
+		idx           string
+	)
+	if wildcard(r.service) || wildcard(r.namespace) {
+		serviceList = k.APIConn.ServiceList()
+	} else {
+		idx = r.service + "." + r.namespace
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go:353]: dnsutil.Join()?

File: plugin/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -357,8 +367,13 @@ func (k *Kubernetes) findServices(r recordRequest, zone string) (services []msg.
 
 		// Endpoint query or headless service
 		if svc.Spec.ClusterIP == api.ClusterIPNone || r.endpoint != "" {
-
-			for _, ep := range k.APIConn.EndpointsList() {
+			if wildcard(r.service) || wildcard(r.namespace) {
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go:370]: Is the check from line 350 now done again?

File: plugin/kubernetes/ns.go [hunk 6]
```diff
@@ -23,7 +23,7 @@ func (k *Kubernetes) nsAddr() *dns.A {
 	rr.A = localIP
 
 FindEndpoint:
-	for _, ep := range k.APIConn.EndpointsList() {
+	for _, ep := range k.APIConn.EpIndexReverse(localIP.String()) {
```
Comment (miekg) [inline plugin/kubernetes/ns.go:26]: Is it an idea to make `EpIndexReverse` take a `net.Addr`?

**Comment by miekg** [comment]
these empty interfaces are not good and set a bad example for code in CoreDNS.
Is anyone volunteering to clean this up? Otherwise I'll probably need to revert.

**Comment by miekg** [comment]
A revert of this passes travis: https://travis-ci.org/coredns/coredns/builds/290356063

**Comment by johnbelamaric** [comment]
Yeah, this is a performance fix so it will pass without it. But the QPS and latency with 10,000 services go from 10k and 0.1ms to 400 and 1000ms without this. Because it is then looping through all (well half on average) services for every query.

I believe the empty interfaces are due to the way we have to interface to the the K8s client-go code. But we need real types so then we have to do type assertions so we can access the data on there.

On Oct 20, 2017, at 4:59 AM, Miek Gieben <notifications@github.com<mailto:notifications@github.com>> wrote:


A revert of this passes travis: https://travis-ci.org/coredns/coredns/builds/290356063

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Plugin/Kubern..." ]
>Yeah, this is a performance fix so it will pass without it. But the QPS and latency with 10,000 services go from 10k and 0.1ms to 400 and 1000ms without this. Because it is then looping through all (well half on average) services for every query.
>
>I believe the empty interfaces are due to the way we have to interface to the the K8s client-go code. But we need real types so then we have to do type assertions so we can access the data on there.

Just checked that. Yes they have an empty interface in there... that's pretty 
terrible

**Comment by miekg** [comment]
Still leaves the failing test:
~~~
== RUN TestKubernetes
--- FAIL: TestKubernetes (3.00s)
	helpers.go:115: rcode is "NXDOMAIN", expected "NOERROR"
	helpers.go:291: ;; opcode: QUERY, status: NXDOMAIN, id: 41324
	;; flags: qr aa rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0
	;; QUESTION SECTION:
	;*.svc-1-a.test-1.svc.cluster.local.	IN	SRV
	;; AUTHORITY SECTION:
	cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1508500831 7200 1800 86400 60
~~~
Suspect we may be faster/slower to start which makes this fail?

**Comment by johnbelamaric** [comment]
Yeah. This is why we need the readiness checks. Think we should have it before 1.0.0

On Oct 20, 2017, at 8:07 AM, Miek Gieben <notifications@github.com<mailto:notifications@github.com>> wrote:


Still leaves the failing test:

== RUN TestKubernetes
--- FAIL: TestKubernetes (3.00s)
        helpers.go:115: rcode is "NXDOMAIN", expected "NOERROR"
        helpers.go:291: ;; opcode: QUERY, status: NXDOMAIN, id: 41324
        ;; flags: qr aa rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0
        ;; QUESTION SECTION:
        ;*.svc-1-a.test-1.svc.cluster.local.    IN      SRV
        ;; AUTHORITY SECTION:
        cluster.local.  300     IN      SOA     ns.dns.cluster.local. hostmaster.cluster.local. 1508500831 7200 1800 86400 60


Suspect we may be faster/slower to start which makes this fail?

**Pull request merged**